pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,230 @@
+package routemetrics
+
+import (
+	"context"
+	"fmt"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a new cache to watch on Route objects from every namespace.
+	newCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cache to the manager so that the cache is started along with the other runnables.
+	mgr.Add(newCache)
+	reconciler := &reconciler{
+		cache:            newCache,
+		Namespace:        namespace,
+		routeToIngresses: make(map[types.NamespacedName]sets.String),
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
```

Comment:
What's the impact on a cluster with thousands of frequently updated routes?  Do we need rate limiting?
```suggestion
	c, err := controller.New(controllerName, mgr, controller.Options{
		Reconciler: reconciler,
		RateLimiter: workqueue.NewMaxOfRateLimiter(
			// Rate-limit to 1 update every 5 seconds per
			// ingresscontroller to avoid burning CPU if route
			// updates are frequent.
			workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 30*time.Second),
			// 10 qps, 100 bucket size, same as DefaultControllerRateLimiter().
			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
		),
	})
```

Reply:
With thousands of routes which are frequently updated, it would make sense to add rate limiting. In fact, the `default` shard will always admit all routes, thus for every route event the reconcile will at least run for the `default` shard. Additionally, if the routes are admitted by other shards as well the reconcile will run for them as well. Thus in a short span of time a lot of ingresscontrollers will be reconciled.

Since, this metric is mainly meant to be sent via telemetry, adding this rate limiting will have negligible effect, if at all, to the telemetry data as the telemetry data is updated every 4mins.

I have added the rate limiting to the controller. I have also run the e2e tests and the rate limiting doesn't have any effect on the wait for the metrics update. The tests passed without any issues.